archive/rpm.ksy

```diff
@@ -0,0 +1,1068 @@
+meta:
+  id: rpm
+  title: RPM Package Manager files
+  file-extension:
+    - rpm
+    - srpm
+    - src.rpm
+    - drpm
+  xref:
+    pronom: fmt/795 # v3
+    wikidata: Q492650
+  license: CC0-1.0
+  ks-version: 0.9
+  endian: be
+doc: |
+  This parser is for the RPM version 3 file format which is the current version
+  of the file format used by RPM 2.1 and later (including RPM version 4.x, which
+  is the current version of the RPM tool). There are historical versions of the
+  RPM file format, as well as a currently abandoned fork (rpm5). These formats
+  are not covered by this specification.
+doc-ref:
+  - https://github.com/rpm-software-management/rpm/blob/911448f2/doc/manual/format.md
+  - https://github.com/rpm-software-management/rpm/blob/911448f2/doc/manual/tags.md
+  - https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/pkgformat.html
+  - http://ftp.rpm.org/max-rpm/
+seq:
+  - id: lead
+    type: lead
+  - id: signature
+    type: header(true)
+  - id: boundary_padding
+    size: (- _io.pos) % 8
+  - id: header
+    type: header(false)
+  # - id: payload
+  #   size: ??
+  #   doc: |
+  #     if signature has a SIZE value, then it is:
+  #     signature[SIZE][0] - sizeof<header>
+types:
+  dummy: {}
+  lead:
+    seq:
+      - id: magic
+        contents: [0xed, 0xab, 0xee, 0xdb]
+      - id: version
+        type: rpm_version
+      - id: type
+        type: u2
+        enum: rpm_types
+      - id: architecture
+        -orig-id: archnum
+        type: u2
+        enum: architectures
+      - id: package_name
+        size: 66
+        type: strz
+        encoding: UTF-8
+      - id: os
+        -orig-id: osnum
+        type: u2
+        enum: operating_systems
+      - id: signature_type
+        type: u2
+        valid: 5
+      - id: reserved
+        size: 16
+  rpm_version:
+    seq:
+      - id: major
+        type: u1
+        valid: 0x3
```

Comment:
> The major version in the header says 3 as well.

I don't know, but the `rpm` software actually allows 4 too ([`lib/rpmlead.c:102-105`](https://github.com/rpm-software-management/rpm/blob/4afe2d14d33db82ccb41c0a8d5eb1a4db90762fc/lib/rpmlead.c#L102-L105)):

```c
    if (lead->major < 3 || lead->major > 4) {
	*msg = xstrdup(_("unsupported RPM package version"));
	return RPMRC_FAIL;
    }
```

But writes 3 ([`lib/rpmlead.c:56`](https://github.com/rpm-software-management/rpm/blob/4afe2d14d33db82ccb41c0a8d5eb1a4db90762fc/lib/rpmlead.c#L56)):

```c
	l->major = 3;
```

Reply:
It has been version 3 for something like 20 years, so I think it is a safe assumption to make. I have never seen a valid file with major version 4.

Reply:
IMHO we should also accept `4`. It may be a plan if RPM developers to make `4` somehow compatible to `3`. Since `rpm` tool accepting `4` is already rolled out, there is no way for rpm devs to reconsider. It would be easier for them just use `5` if they wanted to make incompatible changes.

Reply:
> IMHO we should also accept `4`. It may be a plan if RPM developers to make `4` somehow compatible to `3`. Since `rpm` tool accepting `4` is already rolled out, there is no way for rpm devs to reconsider. It would be easier for them just use `5` if they wanted to make incompatible changes.

I have not seen any version with format version 4. I just went through 21 years of RPM source history (it is all in the RPM Git repository) and it has defaulted to format version 3 ever since and it has never defaulted to 4 at any given time. Note: this is about the RPM version in the `lead` part of the file format.

The reason I put that check in there is for sanity checks during carving.

I would not be opposed to allowing version 4 to be marked as valid, but I just think it is a bit unnecessary, as I am confident the RPM developers will not change it (instead, they will work inside the current format, as they have done in the last 20 years).

Reply:
> I have not seen any version with format version 4. I just went through 21 years of RPM source history (it is all in the RPM Git repository) and it has defaulted to format version 3 ever since and it has never defaulted to 4 at any given time.

Hmm, interestingly enough, I also dug into its Git history and discovered this commit https://github.com/rpm-software-management/rpm/commit/0bac1d421ccdcc719e2e60bcfb4859d0f602510c from 17 Dec 2009 with the following change (apparently reverting back to 3 hardcoded as opposed to switching between 3 and 4 based on some feature added in the past):

```diff
     rpmGetArchInfo(NULL, &archnum);
     rpmGetOsInfo(NULL, &osnum);
 
-    l->major = (_noDirTokens ? 4: 3);
+    l->major = 3;
     l->minor = 0;
     l->archnum = archnum;
     l->osnum = osnum;
```

Change of major version setting to `(_noDirTokens ? 4: 3)` was introduced [in `92698df7`](https://github.com/rpm-software-management/rpm/commit/92698df7236d0fc6c1d9c683210ddc12984764b3#diff-e7052a5203b7bc010c0d314bd59e04f5c7cc17fd4f933465813e3189690ec762R47) (from 1 Dec 2007). So it was in the `rpm` codebase for 2 years (from Dec 2007 to Dec 2009).

Reply:
Folks, the real answer to all this is quite different:
The lead as a structure is 25 years obsolete, the data there is meaningless. Seriously.
Except to check for the magic to detect that it's an rpm file in the first place, just ignore everything in it.